python/cudf/cudf/core/dataframe.py

```diff
@@ -516,6 +516,15 @@ def __setitem__(self, arg, value):
                 if arg in self._data:
                     if is_scalar(value):
                         value = utils.scalar_broadcast_to(value, len(self))
+                    if len(self) == 0:
+                        if isinstance(value, (pd.Series, Series)):
+                            self._index = as_index(value.index)
+                        elif len(value) > 0:
+                            self._index = RangeIndex(start=0, stop=len(value))
+                    elif isinstance(value, (pd.Series, Series)):
+                        value = Series(value)._align_to_index(
+                            self._index, how="right"
+                        )
```

Comment:
This logic is now duplicated in both `__setitem__` and `insert`. Should probably refactor into a function or somehow reduce the duplication between the two.